toolkit/offchain/src/d_param.rs

```diff
@@ -74,109 +75,135 @@ fn get_current_d_parameter(
 	// TODO: await for the transaction output to be visible in validator wallet
 	Ok(McTxHash(res.transaction.id))
 }
 
-async fn update_d_param<C>(
+async fn update_d_param<C: QueryLedgerState + Transactions + QueryNetwork>(
 	validator: &PlutusScript,
 	policy: &PlutusScript,
 	d_parameter: &DParameter,
 	current_utxo: &OgmiosUtxo,
 	ctx: TransactionContext,
+	genesis_utxo: UtxoId,
 	client: &C,
 ) -> anyhow::Result<McTxHash>
 where
 	C: Transactions,
 {
-	let zero_ex_units = ExUnits::new(&0u64.into(), &0u64.into());
-	let tx = update_d_param_tx(validator, policy, d_parameter, current_utxo, &ctx, zero_ex_units)?;
+	let zero_ex_units = ScriptExUnits {
+		mint_ex_units: vec![ExUnits::new(&0u64.into(), &0u64.into())],
+		spend_ex_units: vec![ExUnits::new(&0u64.into(), &0u64.into())],
+	};
+	let tx = update_d_param_tx(validator, policy, d_parameter, current_utxo, &ctx, zero_ex_units, genesis_utxo, client).await?;
 	let evaluate_response = client.evaluate_transaction(&tx.to_bytes()).await.map_err(|e| {
 		anyhow!(
 			"Evaluate update D-parameter transaction request failed: {}, bytes: {}",
 			e,
 			hex::encode(tx.to_bytes())
 		)
 	})?;
-	let spend_ex_units = get_first_validator_budget(evaluate_response)?;
+	let spend_ex_units = get_validator_budgets(evaluate_response)?;
 
-	let tx = update_d_param_tx(validator, policy, d_parameter, current_utxo, &ctx, spend_ex_units)?;
+	let tx = update_d_param_tx(validator, policy, d_parameter, current_utxo, &ctx, spend_ex_units, genesis_utxo, client).await?;
 	let signed_tx = ctx.sign(&tx).to_bytes();
 	let res = client.submit_transaction(&signed_tx).await.map_err(|e| {
 		anyhow!(
 			"Submit D-parameter update transaction request failed: {}, bytes: {}",
 			e,
-			hex::encode(tx.to_bytes())
+			hex::encode(signed_tx)
 		)
 	})?;
 	log::info!("Update D-parameter transaction submitted: {}", hex::encode(res.transaction.id));
 	// TODO: await for the transaction output to be visible in validator wallet
 	Ok(McTxHash(res.transaction.id))
 }
 
-fn mint_d_param_token_tx(
+async fn mint_d_param_token_tx<C: QueryLedgerState + Transactions + QueryNetwork>(
 	validator: &PlutusScript,
 	policy: &PlutusScript,
 	d_parameter: &DParameter,
 	ctx: &TransactionContext,
-	mint_witness_ex_units: ExUnits,
+	mut ex_units: ScriptExUnits,
+	genesis_utxo: UtxoId,
+	client: &C,
 ) -> Result<Transaction, JsError> {
 	let mut tx_builder = TransactionBuilder::new(&get_builder_config(ctx)?);
 	// The essence of transaction: mint token and set output with it
-	tx_builder.add_mint_one_script_token(policy, mint_witness_ex_units)?;
+	tx_builder.add_mint_one_script_token(policy, ex_units.pop_mint_ex_units())?;
 	tx_builder.add_output_with_one_script_token(
 		validator,
 		policy,
 		&d_parameter_to_plutus_data(d_parameter),
 		ctx,
 	)?;
 
+	let gov_utxo = crate::init_governance::get_governance_utxo(genesis_utxo, client).await.map_err(|e| JsError::from_str(e.to_string().as_str()))?;
```

Comment:
This function was "clean sync" before by design and all "dirty async" ogmios async things happen outside and pass parameters into it. Please pass `goveranance_utxo` instead of passing `genesis_utxo` to keep this function clean.